ldap/servers/slapd/daemon.c

```diff
@@ -775,22 +794,88 @@ disk_monitoring_thread(void *nothing __attribute__((unused)))
         return;
 }
 
+char *epoll_event_flags_to_string(PRUint32 events)
+{
+    static char buf[64];
+    int len = 0;
+
+    if (events & EPOLLIN) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLIN|");
+    }
+    if (events & EPOLLOUT) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLOUT|");
+    }
+    if (events & EPOLLERR) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLERR|");
+    }
+    if (events & EPOLLHUP) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLHUP|");
+    }
+    if (events & EPOLLRDHUP) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLRDHUP|");
+    }
+    if (events & EPOLLEXCLUSIVE) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLEXCLUSIVE|");
+    }
+    if (events & EPOLLPRI) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLPRI|");
+    }
+    if (events & EPOLLONESHOT) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLONESHOT|");
+    }
+    if (events & EPOLLET) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLET|");
+    }
+    if (len == 0) {
+        len += snprintf(buf + len, sizeof(buf) - len, "EPOLLUNKNOWN|");
+    }
+    /* Remove the trailing | */
+    if (len > 0 && buf[len - 1] == '|') {
+        buf[len - 1] = '\0';
+    }
+    /* If no flags were set, return a specific message */
+    if (len == 0 || (len == 1 && buf[0] == '\0')) {
+        snprintf(buf, sizeof(buf), "No flags");
+    }
+    return buf;
+}
+
 static void
+#ifdef ENABLE_EPOLL
+handle_listeners(struct epoll_event *events, int event_count)
+#else
 handle_listeners(struct POLL_STRUCT *fds)
+#endif /* ENABLE_EPOLL */
 {
     Connection_Table *ct = the_connection_table;
     size_t idx;
     int ctlist = 0;
+#ifdef ENABLE_EPOLL
+    struct listener_info *listener = NULL;
+    for (idx = 0; idx < (size_t)event_count; idx++) {
+        listener = (struct listener_info *)events[idx].data.ptr;
+        PRFileDesc *listenfd = (PRFileDesc *)listener->listenfd;
+        int secure = listener->secure;
+        int local = listener->local;
+#else /* !ENABLE_EPOLL */
     for (idx = 0; idx < listeners; ++idx) {
         int fdidx = listener_idxs[idx].idx;
         PRFileDesc *listenfd = listener_idxs[idx].listenfd;
         int secure = listener_idxs[idx].secure;
         int local = listener_idxs[idx].local;
+#endif /* ENABLE_EPOLL */
         if (listenfd) {
+#ifdef ENABLE_EPOLL
+            if (events[idx].events & EPOLLIN) {
+#else /* !ENABLE_EPOLL */
             PR_ASSERT(fds != NULL);
             PR_ASSERT(listenfd == fds[fdidx].fd);
             if (SLAPD_POLL_LISTEN_READY(fds[fdidx].out_flags)) {
-                /* accept() the new connection, put it on the active list for handle_pr_read_ready */
+#endif /* ENABLE_EPOLL */
+                /* Question: If the listenfd is non-blocking, shouldn't we loop here until
```

Comment:
the return here can also be -1 if PR_Accept fails, so we don't want to loop here if that already happened.

Generally if we get to this situation we really are at an edge case where we have filled the connection table, so erroring faster is better than looping, because if we loop then we also cause other connections to be delayed in finding out they can't be accepted.